### DISCUSSION TITLE
Select dynamic options default not working

### DISCUSSION THREAD

**Discussion by FXNeoCube** [alpinejs/alpine#2697] [1. Help]
If you set dynamic parameters, the preset values do not work.
"Orange" is output, but "Red" is specified in Select
How do I make it work? Or does Alpine JS not support this?

```
<div x-data="{color: 'Orange'}">
    <select x-model="color" >
        <template x-for="color in ['Red', 'Orange', 'Yellow']">
            <option x-text="color"></option>
        </template>
    </select>
    Colors: <span x-text="color"></span>
</div>
```

**Comment by FXNeoCube** [comment]
The problem is solved.

```
<div x-data="{selected: 'Orange'}">
    <select x-model="selected" >
        <template x-for="color in ['Red', 'Orange', 'Yellow']">
                <option x-text="color" :selected="color == selected"></option>
        </template>
    </select>
    Colors: <span x-text="selected"></span>
</div>
```

**Comment by docdunning** [comment]
`:selected="color == selected"`

Yes, but to my mind, that's a workaround.  Alpine's handing of `x-model` in a `select` is inconsistent :-(

If the `select` is hardwired in the HTML, then `x-model` works OK and the default value is properly selected.

If the `select` options are created in an `x-for` loop, then none of them is pre-selected.  The control and its `x-model` are immediately out of sync.

See the demo page below.  Both selects should have the 'Red' option selected.  The first `select` is created dynamically, and shows 'Orange' (the first option). The second select is hardcoded, and works as you would expect, with 'Red' selected.

This feels like a defect to me, given that the behaviour is just not what you would expect, but I'll see what other people think ...

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="{
  color: 'Yellow',
  color2: 'Red',
}">
<p>
  <select x-model="color">
    <template x-for="c in ['Red', 'Orange', 'Yellow']" :key="c">
        <option :value="c"><span x-text="c"></span></option>
    </template>
  </select>
  Color: <span x-text="color"></span>
</p>
<p>
  <select x-model="color2">
    <option value="Orange">Orange</option>
    <option value="Yellow">Yellow</option>
    <option value="Red">Red</option>
  </select>
  Color2: <span x-text="color2"></span>
</p>
</div>
</body>
</html>

```

**Reply by ekwoka** [reply]
I'm not sure to what degree this can be handled in Alpine...

as it's related to the behavior of the element itself. It doesn't let you set a value that it doesn't have an option for. I guess, at most we could queue the value setting to a microtask on select elements?

**Reply by docdunning** [reply]
> It doesn't let you set a value that it doesn't have an option for

Yes, that makes sense.  The problem is only on the initial pass - after that everything is fine.

The `:selected="color == selected"` approach taken by @FXNeoCube works OK, but it just feels a little untidy to be continually updating the DOM attribute when the bound value changes.

Here's an alternative, using the `select`'s `x-init` event with `nextTick`.  I prefer this because it makes it a one-time process. Maybe this is the same as your suggestion for a microtask - although that's beyond my experience.

```
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <title>Document</title>
  <script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.13.5/dist/cdn.min.js"></script>
</head>
<body>
<div x-data="{
  color: 'Red',
}">
<p>
  <select x-model="color" x-init="$nextTick(() => $el.value = color )">
    <template x-for="c in ['Orange', 'Yellow', 'Red']" :key="c">
        <option :value="c" x-text="c"></option>
    </template>
  </select>
  Color: <span x-text="color"></span>
</p>
</div>
</body>
</html>
```

**Reply by ekwoka** [reply]
yes. nextTick just does it on the next frame. similar thing.

**Reply by SimoTod** [reply]
I suppose someone could just send a PR for x-model. If tag is select set the value with a micro task or something like that

**Comment by ljk0071** [comment]
For anyone struggling with cascading selects not updating x-model, here is a "set and forget" solution. It uses a custom directive to watch for DOM changes in the options and forces a refresh.

       document.addEventListener('alpine:init', () => {
            Alpine.directive('watch-options', (el, {}, {cleanup, evaluate}) => {
                const observer = new MutationObserver(() => {
                    const modelName = el.getAttribute('x-model');
                    if (modelName) {
                        const currentValue = evaluate(modelName);
                        const values = Array.from(el.options).map(opt => opt.value);

                        if (!values.includes(currentValue)) el.dispatchEvent(new Event('change', {bubbles: true}))
                        else if (el.value !== currentValue) el.value = currentValue
                    }
                });

                observer.observe(el, {childList: true, subtree: true});

                cleanup(() => observer.disconnect());
            });
        });

**Comment by ivangusev** [comment]
Triggering change doesn't solve the problem when **x-for** is used. Here is correctly working solution:
```
Alpine.directive('monitor-options', (el, { expression }, { evaluate, cleanup }) => {
        const observer = new MutationObserver(() => {
            // 1. Get the name of the variable bound to x-model (e.g., "selectedValue")
            const modelName = el.getAttribute('x-model');
            
            // 2. If x-model exists, grab the current value from Alpine data
            if (modelName) {
                const currentValue = evaluate(modelName);
                
                // 3. Force the select box to match that value immediately
                // This overrides the browser trying to select the first option
                if (currentValue !== undefined) {
                    el.value = currentValue;
                }
            }
        });

        observer.observe(el, { childList: true, subtree: true });

        cleanup(() => observer.disconnect());
    });
```

**Comment by ljk0071** [comment]
i wonder what case u solve problem using by that code.

in my case, ur code is not working.

<select id="region-select"
        x-model="region"
        x-monitor-options>
    <template x-for="r in regions" :key="r.id">
        <option :value="r.id" x-text="r.name"></option>
    </template>
</select>

**Reply by ivangusev** [reply]
if you populate `<select x-model="selectedValue" monitor-option>` options like `<template x-for="option in options">` and want **selectedValue** to be shown, then you need to "select" corresponding `<option>`.
my code above does exactly this

**Reply by ljk0071** [reply]
I think there may be a misunderstanding between us.

The code you wrote changes the visible options based on the value bound to x-model on the select element. That case can already be handled with something like :selected="r.id === region".

The code I proposed, however, is intended for a different scenario: when there are multiple select boxes and the value of the first select box changes the options of the second one. In that case, the goal is to ensure that the x-model value of the second select box follows (is updated to) the value of the newly changed options.

**Reply by ivangusev** [reply]
> there are multiple select boxes and the value of the first select box changes the options of the second one

If you are talking about **initial** value of first select box then still no need to trigger change event. Here is example of second select box:
`<template x-for="state in availableStates">`
where **availableStates** is a function like:
```
get availableStates() {
    return this.countries[this.selectedValue];
},
```
This works fine because **selectedValue** you set in **x-data**.
While my code above for **monitor-options** you would need if you want also to set **initial** value for second select box.

> That case can already be handled with something like :selected="r.id === region".

Yes, but it's a trick, while **monitor-options** is a fix of alpinejs timing issue.

**Reply by ljk0071** [reply]
Hmm… what I’m talking about is this case:

When the value of the first select box changes, the options of the second select box are updated.
At that point, if the value bound to the second select box via x-model does not exist in the newly updated options, none of the options becomes selected.

In this situation, what the user sees are the new options, but the x-model value does not follow any of the options currently rendered on the page.
So visually the options have changed, but the state held by x-model is out of sync with what’s actually displayed.

**Reply by ivangusev** [reply]
Just do like this:
```
<select x-model="selectedValue" @change="onCountryChange()"
onCountryChange() {
            this.state = ''; // reset value of second select box
        },
```

**Reply by ljk0071** [reply]
i know that is work. but i dont think that is right way

**Reply by ivangusev** [reply]
It seems correct, non-tricky way for cascade select boxes.
